### PR TITLE
Implement PJR checker

### PR DESCRIPTION
**Pull request by coriolinus**
Continuation of #7191. Adds a function which, given the complete set of voting data and a proposed solution, can determine whether or not that solution satisfies _t_-PJR for some selected _t_.

- [x] implement checker
- [x] ~~determine an appropriate "standard" value for _t_ which we can use at least in the tests~~ develop an intuition for what _t_ means, to aid in test scenario construction
- [x] implement fuzz test which ensures that seq-phragmen without fuzzing always passes the checker

### CODE REVIEW

**Comment by kianenigma** [comment]
> determine an appropriate "standard" value for t which we can use at least in the tests develop an intuition for what t means, to aid in test scenario construction

Also, note that this will probably be a configuration in any case. So we don't need to decide on this now and here.

File: primitives/npos-elections/src/pjr.rs [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+/// 3. Voter [`budget`](Voter::budget).
+///
+/// None of the `load` or `score` values are used and can be ignored. This is similar to
+/// [`setup_inputs`] function of this crate.
+///
+/// ### Performance (Weight) Notes
+///
+/// Note that the current function is rather unfortunately inefficient. The most significant
+/// slowdown is the fact that a typical solution that need to be checked for PJR only contains a
+/// subset of the entire NPoS edge graph, encoded as `staked_assignment`. This only encodes the
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:65]:
```suggestion
/// subset of the entire NPoS edge graph, encoded as `supports`. This only encodes the
```

File: primitives/npos-elections/src/pjr.rs [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+	}).collect::<Vec<_>>();
+
+	(candidates, voters)
+}
+
+/// Check a solution to be t-PJR.
+///
+/// ### Semantics
+///
+/// The t-PJR property is defined in the paper "Validator Election in Nominated Proof-of-Stake",
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:157]: reference would be nice here.

Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: agree: 4cff663b0

File: primitives/npos-elections/fuzzer/src/common.rs [hunk 3]
```diff
@@ -39,11 +39,85 @@ pub fn to_range(x: usize, a: usize, b: usize) -> usize {
...
+/// Given parameters governing how many candidates and voters should exist, generates a voting
+/// scenario suitable for fuzz-testing an election algorithm.
+///
+/// The returned candidate list is sorted. This sorting property should not affect the result of the
+/// calculation.
+///
+/// The returned voters list is sorted. This enables binary searching for a particular voter by
+/// account id. This sorting property should not affect the results of the calculation.
+///
+/// Note that this does not generate balancing parameters.
```
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:58]: what does it means 'balancing parameters'?

File: primitives/npos-elections/fuzzer/src/common.rs [hunk 4]
```diff
@@ -39,11 +39,85 @@ pub fn to_range(x: usize, a: usize, b: usize) -> usize {
...
+		let mut id = rng.gen();
+		// insert returns `false` when the value was already present
+		while !used_ids.insert(id) {
+			id = rng.gen();
+		}
+		candidates.push(id);
+	}
+	candidates.sort_unstable();
+	candidates.dedup();
+	debug_assert_eq!(candidates.len(), candidate_count);
```
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:87]: In this context I would assert in all cases, not just debug.

File: primitives/npos-elections/fuzzer/src/common.rs [hunk 5]
```diff
@@ -39,11 +39,85 @@ pub fn to_range(x: usize, a: usize, b: usize) -> usize {
...
+	// cache for fast generation of unique candidate and voter ids
+	let mut used_ids = HashSet::with_capacity(candidate_count + voter_count);
+
+	// always generate a sensible desired number of candidates: elections are uninteresting if we
+	// desire 0 candidates, or a number of candidates >= the actual number of candidates present
+	let rounds = rng.gen_range(1, candidate_count);
+
+	// candidates are easy: just a completely random set of IDs
+	let mut candidates: Vec<AccountId> = Vec::with_capacity(candidate_count);
+	for _ in 0..candidate_count {
```
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:77]: I don't have anything against the approach of using random ids for candidates, but if there are no problems with it, I'd consider using the sorted range as ids, to achieve a simpler code. Example; if you want `x` candidates and `y` votes, candidates will be `0..x` and voters will be `x..x+y`.

File: primitives/npos-elections/fuzzer/src/common.rs [hunk 6]
```diff
@@ -39,11 +39,85 @@ pub fn to_range(x: usize, a: usize, b: usize) -> usize {
...
+		while !used_ids.insert(id) {
+			id = rng.gen();
+		}
+
+		let vote_weight = rng.gen();
+
+		// it's not interesting if a voter chooses 0 or all candidates, so rule those cases out.
+		let n_candidates_chosen = rng.gen_range(1, candidates.len());
+
+		// I believe, but am not 100% certain, that this produces a uniform random distribution of
```
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:102]: Again, consider this as a simpler approach https://docs.rs/rand/0.8.3/rand/seq/trait.SliceRandom.html#tymethod.choose

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		fuzz!(|data: (usize, usize, u64)| {
+			let (mut candidate_count, mut voter_count, seed) = data;
+			let rng = rand::rngs::SmallRng::seed_from_u64(seed);
+			candidate_count = to_range(candidate_count, 100, 1000);
+			voter_count = to_range(voter_count, 100, 2000);
+
+			let (rounds, candidates, voters) =
+				generate_random_npos_inputs(candidate_count, voter_count, rng);
+
+			// Run the NPoS validator selection algorithm
```
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:57]:
```suggestion
			// Run sesq-phragmen.
```

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+			let assignments: Vec<StakedAssignment<AccountId>> = assignments
+				.into_iter()
+				.map(|assignment| {
+					let voter_idx = voters
+						.binary_search_by_key(&assignment.who, |(id, _weight, _assignments)| *id)
+						.expect("voter must be present in voters list");
+					let weight: ExtendedBalance = voters[voter_idx].1.into();
+					assignment.into_staked(weight)
+				})
+				.collect();
```
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:77]: There's a method for this in helpers.rs

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+				winners,
+				assignments,
+			} = seq_phragmen::<AccountId, PerThing>(rounds, candidates.clone(), voters.clone(), None)
+				.expect("seq_phragmen must succeed");
+
+			// pjr_check only cares about the identity of the winner, not its balance
+			let winners: Vec<_> = winners.into_iter().map(|(id, _stake)| id).collect();
+
+			// convert assignments into staked assignments
+			let assignments: Vec<StakedAssignment<AccountId>> = assignments
```
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:68]: there's a method for this in the helpers, if you can build a closure that gives the weight.

File: primitives/npos-elections/fuzzer/src/common.rs (hunk 3 above)
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/common.rs:58]: The `balance` param here: https://github.com/paritytech/substrate/blob/8049fcfa89bfb7375f467a3f22028fea94e7bfe8/primitives/npos-elections/src/phragmen.rs#L66-L71

File: primitives/npos-elections/fuzzer/src/common.rs (hunk 4 above)
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/common.rs]: Sure: 2451a864d

File: primitives/npos-elections/fuzzer/src/common.rs (hunk 6 above)
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/common.rs]: Great idea! 1d4e2dba4

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs (hunk 9 above)
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/phragmen_pjr.rs]: Good idea: fc64ff572

File: primitives/npos-elections/fuzzer/src/common.rs (hunk 3 above)
Comment (kianenigma) [reply primitives/npos-elections/fuzzer/src/common.rs:58]: ahh okay.

**Comment by coriolinus** [comment]
Did some research to determine a small case that fails the fuzzer. This is deterministic: apply the patch in [small_case.diff](https://github.com/paritytech/substrate/files/6034880/small_case.diff.txt), and run the command from the first line of [small_case.txt](https://github.com/paritytech/substrate/files/6034881/small_case.txt), and you'll see roughly the same output. I replaced the candidate and voter IDs with textual identifiers to make life easier, and sorted the voting lists, but that's it for preprocessing. 

In this scenario, there are 12 candidates, 11 winners, and 10 voters. However, none of the voters voted for candidate C or J. Despite that, the algorithm had to pick _someone_ for the 11th winner; it picked C. This appears to me to be a legitimate case where we haven't preserved PJR, because J has a accurate complaint that they had exactly as much voting power allocated to them as C, but didn't make it into the chosen set. 

It's a degenerate case, but then, the point of fuzzing is to discover degenerate cases.

File: primitives/npos-elections/src/pjr.rs [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+///
+/// The t-PJR property is defined in the paper ["Validator Election in Nominated Proof-of-Stake"][NPoS],
+/// section 5, definition 1.
+///
+/// In plain language, the t-PJR condition is: if there is a group of `N` voters
+/// who have `r` common candidates and can afford to support each of them with backing stake `t`
+/// (i.e `sum(stake(v) for v in voters) == r * t`), then this committee needs to be represented by at
+/// least `r` elected candidates.
+///
+/// Section 5 of the NPoS paper shows that this property is equal to: For a feasible solution, if
```
Comment (AlistairStewart) [inline primitives/npos-elections/src/pjr.rs:165]: While the sentence after the colon is correct, the property isn't equal to anything. The test implies the condition, but not every solution that satifies the condition passes the test.

**Comment by AlistairStewart** [comment]
> This appears to me to be a legitimate case where we haven't preserved PJR, because J has a accurate complaint that they had exactly as much voting power allocated to them as C, but didn't make it into the chosen set.

The test should not fail for this. It should ve ok to choose J. Because of rounding errors, it should be ok to choose C even when if J's score is slightly higher than C's support.

Here, the set does not satisfy t-PJR for t=0 but should for any t>0 and the test should show that, because J's score should be 0.

**Comment by coriolinus** [comment]
@AlistairStewart So, the fundamental issue was our unjustified assumption that PJR requires that all elected candidates have non-0 support. 6419d17 removes that requirement. I'll try to find a different failing case now.

File: primitives/npos-elections/src/pjr.rs (hunk 10 above)
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: Updated verbiage in 32da70d.

**Comment by coriolinus** [comment]
I have identified another case in which seq_phragmen produces a result which does not pass PJR. Unfortunately, I couldn't reproduce this case with smaller inputs, which means that this one is moderately large: [failure_medium.txt](https://github.com/paritytech/substrate/files/6042971/failure_medium.txt). (This file used [`symbolicify.py`](https://gist.github.com/coriolinus/d6bfaf490cf1985d99df3cc547fc327c) to generate symbolic representations of the IDs for user friendliness.

In short, the parameters are 87 candidates, 73 winners, 79 voters, and seed 0. The PJR checker says that _t_ is `9 109 055 974 701 609 990`, but Candidate CB has pre-score `11 525 351 653 622 188 319`. 

I don't yet know why `seq_phragmen` produced a non-PJR result here. Running `phragmms` on the same input also produces a result which `pjr_check` rejects.

**Comment by coriolinus** [comment]
Did manage to find a smaller instance of this failure: [fail_medium_2.txt](https://github.com/paritytech/substrate/files/6043466/fail_medium_2.txt)

This one has only 32 candidates, 28 winners, and 5 voters. As before, running the same scenario through `phragmms` also generates a failing winner distribution. This one should be somewhat more tractable to analyze.

**Comment by coriolinus** [comment]
Pulling out the comment from the final commit (as of now), for ease of review:

> It turns out that the fundamental problem causing previous implementations
> of the fuzzer to fail wasn't in `seq_phragmen` _or_ in `pjr_check`: it was
> in the rounding errors introduced in the various conversions between the
> internal data representation and the external one.
> 
> Fixing the fuzzer is then simply an issue of using the internal representation
> and staying in that representation. However, that leaves the issue that
> `seq_phragmen` occasionally produces an output which is technically not
> PJR due to rounding errors. In the future we will need to add some kind of
> "close-enough" threshold. However, that is explicitly out of scope of
> this PR.

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+use common::{generate_random_npos_inputs, to_range};
+use rand::{self, SeedableRng};
+use sp_npos_elections::{pjr_check_core, seq_phragmen_core, setup_inputs, standard_threshold};
+
+type AccountId = u64;
+
+const MIN_CANDIDATES: usize = 5;
+const MAX_CANDIDATES: usize = 200;
+const MIN_VOTERS: usize = 5;
+const MAX_VOTERS: usize = 500;
```
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:58]:
```suggestion
const MIN_CANDIDATES: usize = 500;
const MAX_CANDIDATES: usize = 2000;
const MIN_VOTERS: usize = 1000;
const MAX_VOTERS: usize = 5000;
```

These are more realistic polkadot values, and I think it should not get super slow yet.

File: primitives/npos-elections/src/phragmen.rs [hunk 12]
```diff
@@ -63,6 +63,9 @@ const DEN: ExtendedBalance = ExtendedBalance::max_value();
 /// `expect` this to return `Ok`.
 ///
 /// This can only fail if the normalization fails.
+///
+/// Note that rounding errors can potentially cause the output of this function to fail a PJR check.
```
Comment (kianenigma) [inline primitives/npos-elections/src/phragmen.rs:67]:
```suggestion
/// Note that rounding errors can potentially cause the output of this function to fail a t-PJR check where t is ....
```

better to specify which PJR test exactly

File: primitives/npos-elections/src/pjr.rs [hunk 13]
```diff
@@ -0,0 +1,523 @@
...
+///
+/// A sensible user of this module should make sure that the PJR check is executed and checked as
+/// little as possible, and take sufficient economical measures to ensure that this function cannot
+/// be abused.
+fn prepare_pjr_input<AccountId: IdentifierT>(
+	supports: &Supports<AccountId>,
+	all_candidates: Vec<AccountId>,
+	all_voters: Vec<(AccountId, VoteWeight, Vec<AccountId>)>,
+) -> (Vec<CandidatePtr<AccountId>>, Vec<Voter<AccountId>>) {
+	// collect all candidates and winners into a unified `Vec<Candidate>`.
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:79]:
```suggestion
```

Seems unrelated or outdated.

File: primitives/npos-elections/src/pjr.rs [hunk 14]
```diff
@@ -0,0 +1,523 @@
...
+	// For large lists, that's pretty bad.
+	//
+	// A `SupportMap`, as a `BTreeMap`, has access timing of `O(lg n)`. This means that constructing
+	// the map and then indexing from it gives us timing of `O((s + c) * lg(s))`. If in the future
+	// we get access to a deterministic `HashMap`, we can further improve that to `O(s+c)`.
+	//
+	// However, it does mean allocating sufficient space to store all the data again.
+	let supports: SupportMap<AccountId> = supports.iter().cloned().collect();
+
+	let candidates = all_candidates.into_iter().enumerate().map(|(i, c)| {
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:103]:
```suggestion
	// collect all candidates and winners into a unified `Vec<CandidatePtr>`.
	let candidates = all_candidates.into_iter().enumerate().map(|(i, c)| {
```

File: primitives/npos-elections/src/pjr.rs [hunk 15]
```diff
@@ -0,0 +1,523 @@
...
+				continue;
+			}
+
+			if let Some(idx) = candidates_index.get(&t) {
+				// if this edge is among the assignments, set the weight as well.
+				let weight = assignment_map
+					.get(&v)
+					.and_then(|d| d.iter().find_map(|(x, y)| if x == &t { Some(y) } else { None }))
+					.cloned()
+					.unwrap_or_default();
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:130]: I wonder if this can ever happen? No afaik, but we're being defensive here and it is okay.

File: primitives/npos-elections/src/pjr.rs [hunk 16]
```diff
@@ -0,0 +1,523 @@
...
+///
+/// In plain language, the t-PJR condition is: if there is a group of `N` voters
+/// who have `r` common candidates and can afford to support each of them with backing stake `t`
+/// (i.e `sum(stake(v) for v in voters) == r * t`), then this committee needs to be represented by at
+/// least `r` elected candidates.
+///
+/// Section 5 of the NPoS paper shows that this property can be tested by: for a feasible solution, if
+/// `Max {score(c)} < t` where c is every unelected candidate, then this solution is t-PJR. However,
+/// this test is incomplete: while every solution which passes this test satisfies t-PJR, not every
+/// solution which is t-PJR passes this test. We therefore look to a more accurate test.
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:163]:
```suggestion
/// `Max {score(c)} < t` where c is every unelected candidate, then this solution is t-PJR.
```

I think this whole sentence is misleading and either need to be reworked or removed. I know that you are trying to address the accuracy issues. But you are phrasing it in a way as if somehow, sometimes, a PJR solution is suddenly no longer PJR. A better assumption is: If a solution `S` is PJR and is no longer PJR due to accuracy issues, it is simply because it is no longer `S` but rather `S'`, a variant thereof.

File: primitives/npos-elections/src/pjr.rs [hunk 17]
```diff
@@ -0,0 +1,523 @@
...
+//
+// Note that while the names don't explicitly say so, `candidates` are the winning candidates, and
+// `voters` is the set of weighted edges from nominators to winning validators.
+pub fn t_pjr_check<AccountId: IdentifierT>(
+	supports: &Supports<AccountId>,
+	all_candidates: Vec<AccountId>,
+	all_voters: Vec<(AccountId, VoteWeight, Vec<AccountId>)>,
+	t: Threshold,
+) -> bool {
+	// First order of business: derive `(candidates, voters)` from `supports`
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:195]:
```suggestion
	// First order of business: derive `(candidates, voters)` from `supports`.
```

File: primitives/npos-elections/src/pjr.rs [hunk 18]
```diff
@@ -0,0 +1,523 @@
...
+/// More accurately:
+///
+/// 1. If `c` exactly has `t` backing or less, then we don't generate any slack.
+/// 2. If `c` has more than `t`, then we reduce it to `t`.
+fn slack<AccountId: IdentifierT>(voter: &Voter<AccountId>, t: Threshold) -> ExtendedBalance {
+	let budget = voter.budget;
+	let leftover = voter.edges.iter().fold(Zero::zero(), |acc: ExtendedBalance, edge| {
+		let candidate = edge.candidate.borrow();
+		if candidate.elected {
+			// TODO: using perbill here is just going to cause annoyance, why not just subtract?
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:254]:
```suggestion
```

This is no longer relevant.

File: primitives/npos-elections/src/pjr.rs [hunk 19]
```diff
@@ -0,0 +1,523 @@
...
+				Perbill::one().min(Perbill::from_rational_approximation(t, candidate.backed_stake))
+				* edge.weight;
+			acc.saturating_add(extra)
+		} else {
+			// No slack generated here.
+			acc
+		}
+	});
+
+	// NOTE: candidate for saturating_log_sub()
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:265]:
```suggestion
	// NOTE: candidate for saturating_log_sub(). Defensive-only. 
```

Comment (kianenigma) [reply primitives/npos-elections/src/pjr.rs:265]: (we always comment such incidents with `defensive` keyword)

File: primitives/npos-elections/src/pjr.rs [hunk 20]
```diff
@@ -0,0 +1,523 @@
...
+/// - `committed_size` is the number of winners of the election.
+/// - `weights` is an iterator of voter stakes. If the sum of stakes is already known,
+///   `std::iter::once(sum_of_stakes)` is appropriate here.
+pub fn standard_threshold(
+	committeed_size: usize,
+	weights: impl IntoIterator<Item = ExtendedBalance>,
+) -> Threshold {
+	weights
+		.into_iter()
+		.sum::<Threshold>()
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:287]: please use fold and satuaring add.

**Review by kianenigma — APPROVED** [review]
LGTM modulo a few small nits.

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs (hunk 11 above)
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:58]: Unfortunately, it appears to be pretty slow. On my personal computer, building and executing a single iteration took 159s in release mode. On my Parity Macbook (2019 edition), it took 182s. 

I haven't yet done any real benchmarking to discover whether the slowdown is in case generation or in the `seq_phragmen` algorithm, or both, but I think iterations measured in minutes are too slow for effective fuzzing.

What's clear is that at least one of those two stages is nonlinear in complexity. Cutting your suggested case sizes in half gives a timing of 19.2s on my personal machine, and 22.6s on the MBP. I think that's a better balance of realism and iteration time.

Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/phragmen_pjr.rs]: 9f928e3a1

File: primitives/npos-elections/src/pjr.rs (hunk 16 above)
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs:163]: So, that was in response to https://github.com/paritytech/substrate/pull/8160#discussion_r581943430. I think the actual intent is more like, "There exist solutions which do not pass this test but are nevertheless t-PJR." 

I'll give this some thought and try to clarify.

File: primitives/npos-elections/src/pjr.rs (hunk 20 above)
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: OK: cabce8fd7e7fcb0ae26aa20169ff69c916070a18

File: primitives/npos-elections/src/pjr.rs (hunk 14 above)
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs:103]: 06c0e8af8

File: primitives/npos-elections/src/pjr.rs (hunk 13 above)
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: 06c0e8af8

File: primitives/npos-elections/src/phragmen.rs (hunk 12 above)
Comment (coriolinus) [reply primitives/npos-elections/src/phragmen.rs]: Agree: 9c7fb95c7

File: primitives/npos-elections/src/pjr.rs (hunk 16 above)
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: What do you think of the phrasing introduced in eb943752a?

Comment (kianenigma) [reply primitives/npos-elections/src/pjr.rs]: Better, but I think it is important to clarify why: accuracy errors.

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs (hunk 11 above)
Comment (kianenigma) [reply primitives/npos-elections/fuzzer/src/phragmen_pjr.rs]: Okay no 159 is way too low. Make it something that is feasible on 1 machine. 

I was using the build-host with 128 cores and got a bit spoiled I think :D

File: primitives/npos-elections/src/pjr.rs (hunk 16 above)
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: No, I think it's possible for solutions to fail (in edge cases) for reasons that don't derive from accuracy errors. I'm basing that on this text:

> The test implies the condition, but not every solution that satifies the condition passes the test.

File: primitives/npos-elections/src/pjr.rs [hunk 21]
```diff
@@ -0,0 +1,519 @@
...
+pub fn standard_threshold(
+	committee_size: usize,
+	weights: impl IntoIterator<Item = ExtendedBalance>,
+) -> Threshold {
+	weights
+		.into_iter()
+		.fold(Threshold::zero(), |acc, elem| {
+			acc.saturating_add(elem)
+		})
+	/ committee_size as Threshold
```
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:67]:
```suggestion
	/ committee_size.max(1) as Threshold
```

Overall, always remember that any code that will end up in the wasm runtime must never never ever panic (even in rare/mistake cases like this).

Comment (kianenigma) [reply primitives/npos-elections/src/pjr.rs:67]: (same reason why I asked for sum to change to fold)

File: primitives/npos-elections/fuzzer/src/common.rs [hunk 22]
```diff
@@ -39,11 +39,81 @@ pub fn to_range(x: usize, a: usize, b: usize) -> usize {
...
+	}
+	candidates.sort_unstable();
+	candidates.dedup();
+	assert_eq!(candidates.len(), candidate_count);
+
+	let mut voters = Vec::with_capacity(voter_count);
+	for _ in 0..voter_count {
+		let mut id = rng.gen();
+		// insert returns `false` when the value was already present
+		while !used_ids.insert(id) {
```
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:95]: I wasn't convinced about random ids tbh and changed this in my local testing to a sequence. but your final call 😁

Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/common.rs:95]: I don't think it's that much more expensive to randomize them, and I prefer "more natural" fuzzing inputs when possible.

File: primitives/npos-elections/src/pjr.rs (hunk 21 above)
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs:67]: Good catch!

File: primitives/npos-elections/src/lib.rs [hunk 23]
```diff
@@ -326,6 +335,19 @@ impl<A: IdentifierT> std::fmt::Debug for Voter<A> {
 }
 
 impl<AccountId: IdentifierT> Voter<AccountId> {
+	/// Create a new `Voter`.
+	pub fn new(who: AccountId) -> Self {
+		Self { who, ..Default::default() }
+	}
+
```
Comment (shawntabrizi) [inline primitives/npos-elections/src/lib.rs:342]:
```suggestion
```

File: primitives/npos-elections/src/pjr.rs [hunk 24]
```diff
@@ -0,0 +1,519 @@
+ // This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
```
Comment (shawntabrizi) [inline primitives/npos-elections/src/pjr.rs:3]:
```suggestion
// Copyright (C) 2021 Parity Technologies (UK) Ltd.
```

**Review by shawntabrizi — APPROVED** [review]
looks very clean to me

**Comment by coriolinus** [comment]
bot merge
